File: env/env.go
package env

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"unicode"
)

// GetOS returns a best-effort OS descriptor.
// - On Linux with /etc/os-release: "<id><version_id>" like "ubuntu22.04".
// - On macOS: "darwin<version>" like "darwin14.2.1".
// - On Windows: "windows<version>".
// - Fallback: returns GOOS like "linux"/"darwin"/"windows".
func GetOS() string {
	goos := strings.ToLower(strings.TrimSpace(runtime.GOOS))

	switch goos {
	case "linux":
		if desc := detectLinuxDescriptor(); desc != "" {
			return desc
		}
		return "linux"
	case "windows":
		if v := detectWindowsVersion(); v != "" {
			return "windows" + v
		}
		return "windows"
	case "darwin":
		if v := detectDarwinVersion(); v != "" {
			return "darwin" + v
		}
		return "darwin"
	default:
		return goos
	}
}

func detectLinuxDescriptor() string {
	f, err := os.Open("/etc/os-release")
	if err != nil {
		return ""
	}
	defer f.Close()

	m := map[string]string{}
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		k, v, ok := strings.Cut(line, "=")
		if !ok {
			continue
		}
		m[k] = strings.Trim(v, `"`)
	}
	id := strings.ToLower(strings.TrimSpace(m["ID"]))
	ver := strings.TrimSpace(m["VERSION_ID"])
	if id == "" || ver == "" {
		return ""
	}
	return id + strings.TrimSpace(ver)
}

func detectWindowsVersion() string {
	out, err := exec.Command(
		"powershell",
		"-NoProfile",
		"-NonInteractive",
		"-Command",
		"[System.Environment]::OSVersion.Version.ToString()",
	).Output()
	if err == nil {
		v := strings.TrimSpace(string(out))
		if v != "" {
			return v
		}
	}

	out, err = exec.Command("cmd", "/c", "ver").Output()
	if err != nil {
		return ""
	}
	return extractFirstVersionLikeToken(string(out))
}

func detectDarwinVersion() string {
	out, err := exec.Command("sw_vers", "-productVersion").Output()
	if err == nil {
		v := strings.TrimSpace(string(out))
		if v != "" {
			return v
		}
	}
	out, err = exec.Command("uname", "-r").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func extractFirstVersionLikeToken(s string) string {
	s = strings.TrimSpace(s)
	start := -1
	for i, r := range s {
		if unicode.IsDigit(r) {
			start = i
			break
		}
	}
	if start < 0 {
		return ""
	}
	end := start
	for end < len(s) {
		c := s[end]
		if (c >= '0' && c <= '9') || c == '.' {
			end++
			continue
		}
		break
	}
	if end <= start {
		return ""
	}
	return s[start:end]
}

// GetArch returns current arch with best-effort variant.
// Examples: amd64v1, armv7, arm64v8.
func GetArch() string {
	env, ok := getGoEnvArchVars()
	if ok {
		goarch := strings.ToLower(strings.TrimSpace(env.GOARCH))
		switch goarch {
		case "":
			// fall through to runtime below
		case "amd64":
			goamd64 := strings.ToLower(strings.TrimSpace(env.GOAMD64))
			if strings.HasPrefix(goamd64, "v") {
				return "amd64" + goamd64
			} else {
				fmt.Printf("unexpected GOAMD64 value: %q\n", goamd64)
			}
			return "amd64"
		case "arm":
			goarm := strings.TrimSpace(env.GOARM)
			if goarm != "" {
				return "armv" + goarm
			} else {
				fmt.Println("unexpected GOARM empty value")
			}
			return "arm"
		case "arm64":
			return "arm64v8"
		default:
			return goarch
		}
	}

	goarch := strings.ToLower(strings.TrimSpace(runtime.GOARCH))
	if goarch == "arm64" {
		return "arm64v8"
	}
	return goarch
}

type goEnvArchVars struct {
	GOARCH  string `json:"GOARCH"`
	GOAMD64 string `json:"GOAMD64"`
	GOARM   string `json:"GOARM"`
}

func getGoEnvArchVars() (goEnvArchVars, bool) {
	out, err := exec.Command("go", "env", "-json").Output()
	if err != nil {
		return goEnvArchVars{}, false
	}
	var v goEnvArchVars
	if err := json.Unmarshal(out, &v); err != nil {
		return goEnvArchVars{}, false
	}
	return v, true
}

// GetCompiler returns the Node.js version string.
// Format: "node<major>" (e.g. "node22") from the installed `node --version`.
func GetCompiler() string {
	out, err := exec.Command("node", "--version").Output()
	if err != nil {
		return ""
	}
	// node --version returns something like "v22.11.0"
	ver := strings.TrimSpace(string(out))
	ver = strings.TrimPrefix(ver, "v")
	if ver == "" {
		return ""
	}
	return "node" + ver
}

// GetCompilerMajor returns the Node.js major version string.
// Format: "node<major>" (e.g. "node22").
func GetCompilerMajor() string {
	out, err := exec.Command("node", "--version").Output()
	if err != nil {
		return ""
	}
	ver := strings.TrimSpace(string(out))
	ver = strings.TrimPrefix(ver, "v")
	if ver == "" {
		return ""
	}
	parts := strings.SplitN(ver, ".", 2)
	return "node" + parts[0]
}

File: src/env/env.js
import fs from "node:fs";
import { execFileSync } from "node:child_process";

export function getOS() {
  const platform = process.platform.toLowerCase().trim();

  switch (platform) {
    case "linux":
      return detectLinuxDescriptor() || "linux";
    case "win32": {
      const version = detectWindowsVersion();
      return version ? `windows${version}` : "windows";
    }
    case "darwin": {
      const version = detectDarwinVersion();
      return version ? `darwin${version}` : "darwin";
    }
    default:
      return platform;
  }
}

function detectLinuxDescriptor() {
  let data;
  try {
    data = fs.readFileSync("/etc/os-release", "utf8");
  } catch {
    return "";
  }

  const values = new Map();
  for (const rawLine of data.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) {
      continue;
    }
    const index = line.indexOf("=");
    if (index < 0) {
      continue;
    }
    const key = line.slice(0, index);
    const value = line.slice(index + 1).replace(/^"|"$/g, "");
    values.set(key, value);
  }

  const id = (values.get("ID") ?? "").toLowerCase().trim();
  const version = (values.get("VERSION_ID") ?? "").trim();
  if (!id || !version) {
    return "";
  }
  return id + version.trim();
}

function detectWindowsVersion() {
  try {
    const out = execFileSync(
      "powershell",
      [
        "-NoProfile",
        "-NonInteractive",
        "-Command",
        "[System.Environment]::OSVersion.Version.ToString()",
      ],
      { encoding: "utf8" },
    );
    const version = out.trim();
    if (version) {
      return version;
    }
  } catch {
    // Fall back to cmd below.
  }

  try {
    const out = execFileSync("cmd", ["/c", "ver"], { encoding: "utf8" });
    return extractFirstVersionLikeToken(out);
  } catch {
    return "";
  }
}

function detectDarwinVersion() {
  try {
    const out = execFileSync("sw_vers", ["-productVersion"], { encoding: "utf8" });
    const version = out.trim();
    if (version) {
      return version;
    }
  } catch {
    // Fall back to uname below.
  }

  try {
    return execFileSync("uname", ["-r"], { encoding: "utf8" }).trim();
  } catch {
    return "";
  }
}

function extractFirstVersionLikeToken(value) {
  const match = value.trim().match(/\d[\d.]*/);
  return match ? match[0] : "";
}

export function getArch() {
  switch (process.arch) {
    case "x64":
      return "amd64v1";
    case "arm64":
      return "arm64v8";
    case "arm": {
      const armVersion = process.config?.variables?.arm_version;
      return armVersion ? `armv${armVersion}` : "arm";
    }
    case "ia32":
      return "386";
    default:
      return process.arch.toLowerCase().trim();
  }
}

export function getCompiler() {
  const version = process.version.replace(/^v/, "").trim();
  if (!version) {
    return "";
  }
  return `node${version}`;
}

export function getCompilerMajor() {
  const version = process.version.replace(/^v/, "").trim();
  if (!version) {
    return "";
  }
  return `node${version.split(".", 1)[0]}`;
}
